### PR TITLE
feat: Form disabled

### PR DESCRIPTION
**Pull request by heiyu4585**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

close https://github.com/ant-design/ant-design/issues/30270

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |         Form disabled    |
| 🇨🇳 Chinese |    Form 级别控制输入组件 disabled 能力     |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/config-provider/DisabledContext.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+const DisabledContext = React.createContext<DisabledType>(false);
+
+export interface DisabledContextProps {
+  disabled?: DisabledType;
+  children?: React.ReactNode;
+}
+
+export const DisabledContextProvider: React.FC<DisabledContextProps> = ({ children, disabled }) => (
+  <DisabledContext.Consumer>
```
Comment (zombieJ) [inline components/config-provider/DisabledContext.tsx:13]: 这边改成：
```tsx
const originDisabled = useContext(DisabledContext);

return <DisabledContext.Provider value={disabled || originDisabled}>
```

File: components/config-provider/context.tsx [hunk 2]
```diff
@@ -3,6 +3,7 @@ import defaultRenderEmpty, { RenderEmptyHandler } from './renderEmpty';
 import { Locale } from '../locale-provider';
 import { SizeType } from './SizeContext';
 import { RequiredMark } from '../form/Form';
+import { DisabledType } from 'antd/es/config-provider/DisabledContext';
```
Comment (zombieJ) [inline components/config-provider/context.tsx:6]: 这边引入路径错了

File: components/config-provider/DisabledContext.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+
+export type DisabledType = true | false | undefined;
+
+const DisabledContext = React.createContext<DisabledType>(false);
```
Comment (zombieJ) [inline components/config-provider/DisabledContext.tsx:5]: export

File: components/config-provider/context.tsx [hunk 4]
```diff
@@ -60,6 +61,8 @@ export const ConfigContext = React.createContext<ConfigConsumerProps>({
   renderEmpty: defaultRenderEmpty,
 });
 
+export const DisabledContext = React.createContext<DisabledType>(false);
```
Comment (zombieJ) [inline components/config-provider/context.tsx:64]: 直接用 `components/config-provider/DisabledContext.tsx` 不用在新建一个

File: components/form/demo/advanced-search.md [hunk 5]
```diff
@@ -1,156 +0,0 @@
----
```
Comment (zombieJ) [inline components/form/demo/advanced-search.md:1]: 为什么 Demo 都没了。 。 。

Comment (heiyu4585) [reply components/form/demo/advanced-search.md:1]: > 为什么 Demo 都没了。 。 。

em...这是开发版本,我梳理一个正式版本提下

**Comment by zombieJ** [comment]
> components/~tmp/form-disabled.tsx(20,48): error TS2339: Property 'diabled' does not exist on type 'FormProps<unknown> & { children?: ReactNode; } & { ref?: Ref<FormInstance<unknown>> | undefined; }'.

这是 `demo/disabled.md` 的 ts 错误。

File: components/form/demo/disabled.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  DatePicker,
+  InputNumber,
+  TreeSelect,
+  Switch,
+} from 'antd';
+
+const { RangePicker } = DatePicker;
+const { TextArea } = Input;
+
+type DisabledType = Parameters<typeof Form>[0]['disabled'];
```
[the remark this answers is not in the document]
Comment (heiyu4585) [reply components/form/demo/disabled.md:34]: get

File: components/form/demo/disabled.md [hunk 7]
```diff
@@ -46,12 +45,15 @@ const FormDisabledDemo = () => {
...
-      <Form.Item label="Form disabled" name="disabled">
+      <Form.Item label="Form disabled" name="disabled" valuePropName="checked">
+        <Checkbox>disabled</Checkbox>
+      </Form.Item>
+      <Form.Item label="Radio">
         <Radio.Group>
-          <Radio.Button value="disabled"> true </Radio.Button>
-          <Radio.Button value=""> false </Radio.Button>
+          <Radio value="apple"> apple </Radio>
+          <Radio value="Pear"> Pear </Radio>
```
Comment (afc163) [inline components/form/demo/disabled.md:56]:
```suggestion
          <Radio value="apple">Apple</Radio>
          <Radio value="pear">Pear</Radio>
```

File: components/form/demo/disabled.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (heiyu4585) [reply components/form/demo/disabled.md]: done

File: components/date-picker/generatePicker/generateRangePicker.tsx [hunk 8]
```diff
@@ -65,65 +67,73 @@ export default function generateRangePicker<DateType>(
       const rootPrefixCls = getPrefixCls();
 
       return (
-        <SizeContext.Consumer>
-          {size => {
-            const mergedSize = customizeSize || size;
-
+        <DisabledContext.Consumer>
```
Comment (zombieJ) [inline components/date-picker/generatePicker/generateRangePicker.tsx:70]: 这个 generater 生成 class component 太发指了，得重构成 HOOKS 的版本。 我给 v5 加个 task。

File: components/form/demo/disabled.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  TreeSelect,
+  Switch,
+  Checkbox,
+} from 'antd';
+
+const { RangePicker } = DatePicker;
+const { TextArea } = Input;
+
+const FormDisabledDemo = () => {
+  const [componentDisabled, setComponentDisabled] = useState<boolean>(false);
```
Comment (zombieJ) [inline components/form/demo/disabled.md:36]: 默认为 true，这样 test snapshot 可以正好测试这个

Comment (heiyu4585) [reply components/form/demo/disabled.md]: done

File: components/date-picker/generatePicker/generateRangePicker.tsx (hunk 8 above)
Comment (heiyu4585) [reply components/date-picker/generatePicker/generateRangePicker.tsx:70]: 这个可以给我吗

Comment (heiyu4585) [reply components/date-picker/generatePicker/generateRangePicker.tsx:70]: > 当然，少侠别客气 😁~~
😁

**Review by zombieJ — APPROVED** [review]

**Comment by phoema** [comment]
考虑支持一下预览态吗，就太好了。

**Comment by suica** [comment]
@afc163 @zombieJ  We can use `DisabledContextProvider` to disable a subset of form items. Should we add such usage explicitly to the document? For example, add to the doc of the `Form` component?

**Pull request merged**